core/crypto/src/signature.rs

```diff
@@ -556,6 +558,62 @@ impl<'de> serde::Deserialize<'de> for SecretKey {
 #[derive(Clone)]
 pub struct Secp256K1Signature([u8; 65]);
 
+impl Secp256K1Signature {
+    pub fn check_signature_values(&self, reject_upper: bool) -> bool {
+        let mut r_bytes = [0u8; 32];
+        r_bytes.copy_from_slice(&self.0[0..32]);
+        let r = U256::from(r_bytes);
+
+        let mut s_bytes = [0u8; 32];
+        s_bytes.copy_from_slice(&self.0[32..64]);
+        let s = U256::from(s_bytes);
+
+        let secp256k1_n = U256::from([
+            0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff,
+            0xff, 0xfe, 0xba, 0xae, 0xdc, 0xe6, 0xaf, 0x48, 0xa0, 0x3b, 0xbf, 0xd2, 0x5e, 0x8c,
+            0xd0, 0x36, 0x41, 0x41,
+        ]);
+
+        if reject_upper {
+            let secp256k1_half_n = secp256k1_n / U256::from(2);
+            // Reject upper range of s values (ECDSA malleability)
+            if s > secp256k1_half_n {
+                return false;
+            }
+        }
+
+        r < secp256k1_n && s < secp256k1_n
+    }
+
+    pub fn recover(
+        &self,
+        msg: [u8; 32],
+    ) -> Result<Secp256K1PublicKey, crate::errors::ParseSignatureError> {
+        let recoverable_sig = secp256k1::RecoverableSignature::from_compact(
+            &SECP256K1,
+            &self.0[0..64],
+            secp256k1::RecoveryId::from_i32(i32::from(self.0[64])).unwrap(),
+        )
+        .map_err(|err| crate::errors::ParseSignatureError::InvalidData {
+            error_message: err.to_string(),
+        })?;
+        let msg = Message::from(msg);
+
+        let res = SECP256K1
+            .recover(&msg, &recoverable_sig)
+            .map_err(|err| crate::errors::ParseSignatureError::InvalidData {
+                error_message: err.to_string(),
+            })?
+            .serialize_vec(&SECP256K1, false);
+
+        let pk = Secp256K1PublicKey::try_from(&res[1..65]).map_err(|err| {
+            crate::errors::ParseSignatureError::InvalidData { error_message: err.to_string() }
+        })?;
```

Comment:
This can never fail (any slice of length 64 will be accepted).

Reply:
Thats right. https://github.com/near/nearcore/pull/4380/commits/2091bfafcae395ea2395e4c0f65c6dd21a7843a3